### PR TITLE
add ability to claim expenditure

### LINKED ISSUE #3491: Expenditures: As a user I can make a claim
[Provided by Michał via Contractors project management platform]

## Description
Adds a button that allows the logged in user to claim their funds and handle claim delays. UI only.

## Acceptance Criteria

- A button and claim details that appears for the connected user.
- Showing the state of each of the recipients

## Design

[Figma Link - User being able to claim can claim](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3530%3A345877)
[Figma Link - User being able to claim can claim + multiple tokens](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3526%3A361943)
[Figma Link - User being able to claim nothing to claim](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3526%3A365870)

Possibility to claim@2xe

Possibility to claim@2x-1

Possibility to claim@2x-2

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description
This PR adds possibility to claim funds and displays state of each recipient. UI only.

[Figma Link - User being able to claim can claim](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3530%3A345877)
[Figma Link - User being able to claim can claim + multiple tokens](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3526%3A361943)
[Figma Link - User being able to claim nothing to claim](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3526%3A365870)

**New stuff** ✨

* Claim funds section
Screenshot 2022-07-15 at 10 22 54
* State of each recipient
Screenshot 2022-07-15 at 10 22 45

### Testing

1. Add payment recipient with values.
2. Add in a main title.
3. Click "Submit draft"
4. Click "Lock values"
5. Click "Escrow funds"
6. Click "Release funds"

Resolves  #3491

### CODE REVIEW

**Review by willm30 — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj Hey, I leave a few notes for your attention. 🙂

1. When you don't enter any delay, I would expect the delay to say "None" and there be something to claim immediately. Below, there is nothing to claim and no delay at all is shown.
no-delay

1. There is an icon next to the payments heading in the design that is missing
payments-icon

1. The "completed" button differs from design (although I personally feel your implementation looks fine. Just bringing to your attention in case it wasn't intentional).
completed

1. There are funds that can be claimed now, but the section header says "nothing to be claimed". Also, when there are multiple tokens, the 'value' label hugs the bottom of the container where it should probably either be centered or stay at the top.
claim-now

1. Hours are displayed differently to the design
hrs-expenditures

1. Probably shouldn't be able to enter the same coin more than once
duplicate-coin

**Comment by agnieszkajarosikloj** [comment]
Screenshot 2022-07-21 at 10 34 49
I've used here FormattedRelativeTime component from Format.js and it's default formatting, when the style is set to "short".
Screenshot 2022-07-21 at 10 46 03
Can it stay this way or does it have to be exactly the same as in the design?

**Comment by arrenv** [comment]
> The "completed" button differs from design (although I personally feel your implementation looks fine. Just bringing to your attention in case it wasn't intentional).

This one is not intended to be a button, it is intended to be a tag/pill component. As being completed, there is no more actions to do and thus a button is not required. So, this should be updated to the design.

> There are funds that can be claimed now, but the section header says "nothing to be claimed". Also, when there are multiple tokens, the 'value' label hugs the bottom of the container where it should probably either be centered or stay at the top.

We can have this say "Now" if there is no timeout, so, that it is consistent as a time indicator.

> Hours are displayed differently to the design

This has something that was brought up previously. Can we just make it all long format for now @agnieszkajarosikloj plese, i.e. minutes, hours, days, months.

> Probably shouldn't be able to enter the same coin more than once

This one is not too big of an issue, I don't think we need to do check for this.

**Review by arrenv — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj looks good, thank you.

Could we please have this timer as the lowest time, i.e. this should show `Now`.

9cc199b8-501c-4f94-8212-54dfcd6932a6

Can the padding here be more consistent to match the design:

b5a44834-a9cc-4717-8bbf-d3ae8af3fc88

**Design**

Screenshot 2022-07-21 114040

File: src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        <FormSection appearance={{ border: 'bottom' }}>
+          <div className={styles.row}>
+            <span className={styles.label}>
+              <FormattedMessage {...MSG.totalClaimable} />
+            </span>
+            <div className={styles.valueContainer}>
+              {totalClaimableWithId.map(({ symbol, amount, id }) => (
+                <div className={styles.value} key={id}>
+                  <Numeral
+                    unit={getTokenDecimalsWithFallback(0)}
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx:151]: it should be token.decimals as an argument to call this function

File: src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  buttonAction,
+  buttonText,
+  buttonIsActive,
+  nextClaimableRecipient,
+  totalClaimable,
+  claimableNow,
+  claimed,
+}: Props) => {
+  const { formatMessage } = useIntl();
+  const convertToTokensWithIds = useMemo(
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx:83]: I think this is unnecessary. You can just use nanoid() as key when you call numeral.

File: src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  },
+  now: {
+    id: 'dashboard.ExpenditurePage.Stages.ClaimFunds.now',
+    defaultMessage: 'now',
+  },
+});
+
+export type Token = Record<string, number>;
+
+interface TokenWithId {
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx:57]: We have a type for token `Token` or `AnyToken`. You can remove any fields that wouldn't be used. I made a log and I see all of this for token:
```
{
token:
address: "0x47b3CfC8EeaEae8Ad83845a19112BeA8ca961fb8"
decimals: 18
iconHash: null
id: "0x47b3CfC8EeaEae8Ad83845a19112BeA8ca961fb8"
name: "b"
processedBalances: (4) [{…}, {…}, {…}, {…}]
symbol: "B"
__typename: "Token"
}
```

File: src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  amount: number;
+  id: string;
+}
+
+interface Props {
+  buttonAction?: () => void;
+  buttonText?: string | MessageDescriptor;
+  buttonIsActive: boolean;
+  nextClaimableRecipient?: Recipient;
+  totalClaimable?: Token;
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/ClaimFunds/ClaimFunds.tsx:68]: I think that you should pass an array if there are several tokens/amounts per user (which I can see is possible). Then you won't need any supplementary functions.

File: src/modules/dashboard/components/ExpenditurePage/Stages/Stages.tsx [hunk 5]
```diff
@@ -70,19 +83,35 @@ interface Props {
...
+    return recipients?.map((recipient) => {
+      const token = getRecipientTokens(recipient, colony);
+      return { ...recipient, value: token };
+    });
+  }, [colony, recipients]);
+
+  const {
+    claimableNow,
+    claimed,
+    totalClaimable,
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Stages/Stages.tsx:110]: I think that all of the about should be an array of objects if there are several tokens/amounts per recipient. Then you don't need to create arrays from them later.

**Review by chinins — CHANGES_REQUESTED** [review]
Huge PR - nice work. Some changes needed for the data handling - as per my comments.

**Review by willm30 — APPROVED** [review]
Looking good from my side

**Review by Julianahlk — CHANGES_REQUESTED** [review]
Question, should this cancel expenditure button still be active in the following steps?

Screen Shot 2022-07-29 at 5 25 30 PM
Screen Shot 2022-07-29 at 5 30 09 PM

Also, I noticed when I do not add a value for the first recipient I do get an error message for not adding anything to the input field

Screen Shot 2022-07-29 at 5 35 37 PM

But when I don't add a value to the second recipient I do not see any warning error in the input field so it was difficult to see what I was missing. 

Screen Shot 2022-07-29 at 5 35 55 PM

**Comment by arrenv** [comment]
> Question, should this cancel expenditure button still be active in the following steps?

Yes, cancelling an expenditure can be done all through, even through the Claiming stage. Just not after it has been fully claimed.

**Comment by agnieszkajarosikloj** [comment]
> Also, I noticed when I do not add a value for the first recipient I do get an error message for not adding anything to the input field

By default, the Input component does not display an error message when field isn't touched. I've added setting touched value for this input in Formik, so now error message should be displayed after clicking submit button.

**Review by Julianahlk — APPROVED** [review]
The input field errors work as I expected now. Thank you for making the changes @agnieszkajarosikloj all good from my end now 👍.

File: src/modules/dashboard/components/ExpenditurePage/TimeRelativeShort/TimeRelativeShort.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  updateInterval?: number;
+  /** Datetime to calculate relative/difference from. */
+  value: Date;
+  formatting?: 'short' | 'long' | 'narrow';
+}
+
+const nearestIntervalOf = (value: number, increment: number) =>
+  Math.round(value / increment) * increment;
+
+const displayName = 'TimeRelative';
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/TimeRelativeShort/TimeRelativeShort.tsx:15]: Please, update the display name:
`dashboard.ExpenditurePage.TimeRelativeShort`

File: src/modules/dashboard/components/ExpenditurePage/TimeRelativeShort/TimeRelativeShort.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+  );
+
+  return (
+    <span {...rest}>
+      <FormattedRelativeTime
+        numeric="auto"
+        updateIntervalInSeconds={updateInterval}
+        value={valueInSeconds}
+        // eslint-disable-next-line react/style-prop-object
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/TimeRelativeShort/TimeRelativeShort.tsx:36]: I don't think that there is a need for the `eslint-disable-next-line` here. When i removed it i didn't have any warning,

**Review by chinins — CHANGES_REQUESTED** [review]
Much better now -very nice. There are still some things to clean up. Please, update and comment once it's ready . I'll try to prioritise the review and respond quickly.

**Comment by agnieszkajarosikloj** [comment]
> Much better now -very nice. There are still some things to clean up. Please, update and comment once it's ready . I'll try to prioritise the review and respond quickly.


@chinins I've just added fixes

File: src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx [hunk 8]
```diff
@@ -94,6 +97,32 @@ const MSG = defineMessages({
   },
 });
 
+const initialValues = {
+  expenditure: 'advanced',
```
Comment (chinins) [inline src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx:101]: You need to use the created enum here as well.

**Review by chinins — APPROVED** [review]
Very nice! Just remember to add the enum for the `advanced` string where I left the comment before merging.

**Review by arrenv — APPROVED** [review]
@agnieszkajarosikloj thanks for your hard work on this. Looks good.

**Pull request merged**